src/config/env.c

```diff
@@ -0,0 +1,487 @@
+/* Pi-hole: A black hole for Internet advertisements
+*  (c) 2023 Pi-hole, LLC (https://pi-hole.net)
+*  Network-wide ad blocking via your own hardware.
+*
+*  FTL Engine
+*  Environment-related routines
+*
+*  This file is copyright under the latest version of the EUPL.
+*  Please see LICENSE file for your rights under this license. */
+
+#include "env.h"
+#include "log.h"
+#include "config/config.h"
+// get_refresh_hostnames_str()
+#include "datastructure.h"
+//set_and_check_password()
+#include "config/password.h"
+// cli_tick()
+#include "args.h"
+// suggest_closest()
+#include "config/levenshtein.h"
+struct env_item
+{
+	bool used;
+	bool valid;
+	char *key;
+	char *value;
+	struct env_item *next;
+};
+
+static struct env_item *env_list = NULL;
+static const char *env_keys[sizeof(config) / sizeof(struct conf_item)] = { NULL };
+
+void getEnvVars(void)
+{
+	// Read environment variables only once
+	if(env_list != NULL)
+		return;
+
+	// Get all environment variables
+	for(char **env = environ; *env != NULL; env++)
+	{
+		// Check if this is a FTLCONF_ variable
+		if(strncmp(*env, FTLCONF_PREFIX, sizeof(FTLCONF_PREFIX) - 1) == 0)
+		{
+			// Split key and value
+			char *key = strtok(*env, "=");
+			char *value = strtok(NULL, "=");
+
+			// Add to list
+			struct env_item *new_item = calloc(1, sizeof(struct env_item));
+			new_item->used = false;
+			new_item->key = strdup(key);
+			new_item->value = strdup(value);
+			new_item->next = env_list;
+			env_list = new_item;
+		}
+	}
+
+	// Add all config item env keys to the list
+	for(unsigned int i = 0; i < sizeof(env_keys) / sizeof(*env_keys); i++)
+	{
+		struct conf_item *conf_item = get_conf_item(&config, i);
+		env_keys[i] = conf_item->e;
+	}
+
+}
+
+void printFTLenv(void)
+{
+	// Nothing to print if no env vars are used
+	if(env_list == NULL)
+		return;
+
+	// Count number of used and ignored env vars
+	unsigned int used = 0, invalid = 0, ignored = 0;
+	for(struct env_item *item = env_list; item != NULL; item = item->next)
+	{
+		if(item->used)
+			if(item->valid)
+				used++;
+			else
+				invalid++;
+		else
+			ignored++;
+	}
+
+	const unsigned int sum = used + invalid + ignored;
+	log_info("%u FTLCONF environment variable%s found (%u used, %u invalid, %u ignored)",
+	         sum, sum == 1 ? "" : "s", used, invalid, ignored);
+
+	// Iterate over all known FTLCONF environment variables
+	for(struct env_item *item = env_list; item != NULL; item = item->next)
+	{
+		if(item->used)
+		{
+			if(item->valid)
+				log_info("%s %s is used", cli_tick(), item->key);
+			else
+				log_info("%s %s is invalid, using default", cli_cross(), item->key);
+			continue;
+		}
+		// else: print warning
+		const char *closest_match = suggest_closest(env_keys, sizeof(env_keys) / sizeof(*env_keys), item->key);
+		log_info("%s %s is unknown and ignored, did you mean %s ?", cli_qst(), item->key, closest_match);
```

Comment:
There are several places where warnings should be logged, but `info` is done.

Reply:
I did this intentionally so the text is more readable, compare:
```
2023-10-31 19:20:03.366 [799762M] INFO: 6 FTLCONF environment variables found (1 used, 1 invalid, 4 ignored)
2023-10-31 19:20:03.367 [799762M] INFO: [?] FTLCONF_debug_ApI is unknown and ignored, did you mean FTLCONF_debug_api ?
2023-10-31 19:20:03.367 [799762M] INFO: [?] FTLCONF_dns_upstrrrr is unknown and ignored, did you mean FTLCONF_dns_upstreams ?
2023-10-31 19:20:03.367 [799762M] INFO: [✗] FTLCONF_debug_api is invalid, using default
2023-10-31 19:20:03.368 [799762M] INFO: [?] FTLCONF_dns_prt is unknown and ignored, did you mean FTLCONF_dns_port ?
2023-10-31 19:20:03.368 [799762M] INFO: [?] FTLCONF_misc_privacyllll is unknown and ignored, did you mean FTLCONF_misc_privacylevel ?
2023-10-31 19:20:03.368 [799762M] INFO: [✓] FTLCONF_misc_nice is used
```
vs.
```
2023-10-31 19:20:03.366 [799762M] INFO: 6 FTLCONF environment variables found (1 used, 1 invalid, 4 ignored)
2023-10-31 19:20:03.367 [799762M] WARNING: [?] FTLCONF_debug_ApI is unknown and ignored, did you mean FTLCONF_debug_api ?
2023-10-31 19:20:03.367 [799762M] WARNING: [?] FTLCONF_dns_upstrrrr is unknown and ignored, did you mean FTLCONF_dns_upstreams ?
2023-10-31 19:20:03.367 [799762M] ERROR: [✗] FTLCONF_debug_api is invalid, using default
2023-10-31 19:20:03.368 [799762M] WARNING: [?] FTLCONF_dns_prt is unknown and ignored, did you mean FTLCONF_dns_port ?
2023-10-31 19:20:03.368 [799762M] WARNING: [?] FTLCONF_misc_privacyllll is unknown and ignored, did you mean FTLCONF_misc_privacylevel ?
2023-10-31 19:20:03.368 [799762M] INFO: [✓] FTLCONF_misc_nice is used
```

We could space-align to make it

vs.
```
2023-10-31 19:20:03.366 [799762M] INFO:     6 FTLCONF environment variables found (1 used, 1 invalid, 4 ignored)
2023-10-31 19:20:03.367 [799762M] WARNING: [?] FTLCONF_debug_ApI is unknown and ignored, did you mean FTLCONF_debug_api ?
2023-10-31 19:20:03.367 [799762M] WARNING: [?] FTLCONF_dns_upstrrrr is unknown and ignored, did you mean FTLCONF_dns_upstreams ?
2023-10-31 19:20:03.367 [799762M] ERROR:   [✗] FTLCONF_debug_api is invalid, using default
2023-10-31 19:20:03.368 [799762M] WARNING: [?] FTLCONF_dns_prt is unknown and ignored, did you mean FTLCONF_dns_port ?
2023-10-31 19:20:03.368 [799762M] WARNING: [?] FTLCONF_misc_privacyllll is unknown and ignored, did you mean FTLCONF_misc_privacylevel ?
2023-10-31 19:20:03.368 [799762M] INFO:    [✓] FTLCONF_misc_nice is used
```